data/contract_docs/logs.md

```diff
@@ -0,0 +1,97 @@
+# Logs
+
+Status: EARLY Draft/Proposal
+
+## LogLines
+
+### Properties and field requirements
+
+- **Time field** - _required_
+  - There must be at least one non nullable time field
+  - If there are multiple time fields present, following will decide the priority
+    - First matching time field with name `timestamp`
+    - or first matching time field with name `ts`
+    - or first matching time field
+- **Message field** - _required_
+  - There must be at lease one non nullable string field must present
+  - If more than one string fields found, the following will decide the priority
+    - First matching string field with name `body`
+    - or first matching string field with name `message`
+    - or first matching string field
+- **Severity field** - _optional_
+  - This is optional field
+  - Level/Severity of the log line can be represented with this field.
+  - This have to be a string field. (either nullable or non-nullable string field)
+  - Log field will be decided in the following order
+    - First matching string field with name `severity`
+    - or first matching string field with name `level`
+  - If no level field found, consumers/client will decide the log level. Example: logs panels will try to parse the message field and determine the log level
+  - Log level can be one of the values specified in the docs [here](https://grafana.com/docs/grafana/latest/explore/logs-integration/)
+- **ID field** - _optional_
+  - This optional field
+  - Unique identified of the log line
+  - This have to be a string field. (either nullable or non-nullable string field)
+  - Id field will be decided in the following order
+    - First matching string field with name `id`
+    - or first matching string field with name `guid`
+  - If no id field found, consumers/client will decide the id field as required.
+- **Attributes field** - _optional_ / **Labels field**
+  - This is an optional field
+  - This field represent additional attributes of the log line. This is also known as labels field.
+  - Field type must be json raw message type. Example value: `{}`, `{"hello":"world", "foo": 123.45, "bar" :["yellow","red"], "baz" : { "name": "alice" }}`
+    - Should not be empty string.
+    - Value should be represented with `Record<string|number,any>` type in javascript.
+  - Attribute field will be decided in the following order
+    - First matching string field with name `attributes`
+    - or first matching string field with name `labels`
+
+### Additional fields
+
+- **NanoSecond Time field** - _optional_
+
+  - When the log line have sub-milli second precisions, regular time field not suitable to represent them.
+  - Field type must be non-nullable string and field name must be `tsNs`
+  - when this field detected, clients are suggested to use this field and ignore the main time field
+  - Field values can't be empty and must have only positive numbers. (no decimal places or floating numbers)
```

Comment:
random idea: what if this field becomes a "javascript number", where we store the "lost precision" from the nanosecond->millisecond transform.
in other words, if you have this nanosecond timestamp: 123456789, you would store it like this:
- timestamp-field: 123
- nanosec-field: 456789
what do you think?

Reply:
in our last chat, the idea of a new field-type was talked about, that we could use as a single timestamp field,  for example it could contain a pair of numbers, where the first number is the millisecond-timestamp, and the second number is the "nanosecond offset",a number between `0` and `999999`

pinging @leeoniya to talk about performance considerations or alternative approaches.

Reply:
> yes it'll waste space

actually, we have a way to encode this efficiently over the wire, so should be okay:

yes it'll waste space

Reply:
@leeoniya 

thanks for all the info 👍 

> i think the sdk now has a `timeOffset` field type for this. so we can require a `time` OR `timeOffset` field. that doesnt feel great tho, and there's currently no support for rendering timeOffset anywhere.

hmmm.. the description says the "timeOffset" is "from the start" is that the "start-of-dataframe"? can it be "from the timestamp-field-in-the-same-row"? 😄 ... how does it work when you have data with a large gap in the "middle"?

- `line1`
- (1ns gap) 
- `line2`
- (1ns gap)
- `line3`
- (1year gap) <---
- `line4`
- (1ns gap)
- `line5`

(regarding support for displaying it: these special timestamps will be used primarily with logs (at least at first i think), and we can adjust the logs-panel to handle them) 
> 
> i think js numbers can support enough float precision to just stick the ns portion after the ms timestamp in normal `time` fields (yes it'll waste space and you'll get precision errors at some 30th decimal place). we can easily probe the data to figure out if we have integers (ms timestamps) or floats (ms.ns timestamps). this gets us mostly transparent support for existing vis, and we can gradually introduce detection and swap rendering & display strategies. maybe.

my problem with this approach is that it does not get us to full nanosecond "precision". technically at one point we only had millisecond-precision and it was not enough, so we went to nanoseconds (Loki uses nanoseconds). this approach would reduce the precision somewhat (to microseconds perhaps? too lazy to do the math 😄 )... so, if we go with this, we need to accept that we are making the quality worse than what it is right now (it might be the right decision, but we have to explicitly decide so)

Reply:
> this approach would reduce the precision somewhat

you're right, i thought we had more decimals to work with than we actually do. looks like epoch timestamp over 1us resolution would have to be a string :sob: 

![image](https://user-images.githubusercontent.com/43234/214648992-85c3478e-9393-4692-9d4d-e20e2ae9eda1.png)

> hmmm.. the description says the "timeOffset" is "from the start" is that the "start-of-dataframe"? can it be "from the timestamp-field-in-the-same-row"?

https://github.com/grafana/grafana-plugin-sdk-go/pull/589/files#diff-6bb11b14801c6bebc0f28f2373ecbe8b4d6f81f5dcaa855371b790b1b4708f34R87

`field.config.type.timeOffset.startTime` + `nanoOffset` is the start nano time. the field values are all in `offsetUnit` units from that.

> how does it work when you have data with a large gap in the "middle"?

with undefined behavior, of course!

imo, except for a table or raw log lines vis when you have from/to that exceeds the available pixels, you can shed many orders of magnitude in precision from the response [by min-clamping the step/interval] without any meaningful effect on interpreting the data. some kind of agg/summary/downsampling + drill-down is unavoidable when you have 3e6 raw log lines in the query range -- whether it's a year range with millis, or an hour range with nanos.

Reply:
@leeoniya wrote:
> e.g. what does influx do when you ask for a year's worth of nanos? and how important are nanos really when you have a year span? even 1s seems like extreme overkill. the same is true of your example of ns, ns, ns, YEAR, ns.

(i assume you mean Loki here, not Influx 😄 .. still, Influx can do nanos too, the result-format is CSV and the timestamps will be in the `2023-01-26T07:52:42.12345Z` format)

i think the difference here is that i look at this from the "display logs" perspective and you from the "display timeseries numbers" perspective. for logs, if you ask for a year's worth of logs, you get all those logs (the database has a limit on how many log-lines it is willing to return at a single time, so that applies).

i understand that when you want to display those timestamps-with-numbers in a TimeSeries panel, then nanoseconds+large_gaps will not make much sense.
(also, usually you group-by-interval, so for a large-time-range the "step" will be quite large anyway, so nanoseconds are not important.. meaning the nanoseconds are only important when you visualize a small time range without large gaps)

Reply:
seems we have an agreement on how to approach this ( https://github.com/grafana/grafana-plugin-sdk-go/pull/647 ), i'm marking this part of the conversation resolved.